### DISCUSSION TITLE
websockify on android

### DISCUSSION THREAD

**Discussion by CorruptionVR** [yume-chan/ya-webadb#512] [Q&A] [answered]
hey, is it possible to start and connect the websockify server on the android device with termux or something?

**Comment by yume-chan** [comment, marked as answer]
Termux has Node.js, I believe WebSockify-JS will run on Android. Just note that WebSocket connections to other machines must use TLS encryption. Running both Tango and WebSockify on same Android device should work like running them on same PC.

**Reply by CorruptionVR** [reply]
Can I output that to a wss to use on another computer?

**Reply by yume-chan** [reply]
Yes, if you have a valid certificate or trust your self-generated certificate on that computer. See #474

**Reply by parkerlreed** [reply]
Yeah works fine within Node JS in Termux. Done that on a few devices.

**Reply by CorruptionVR** [reply]
@parkerlreed, could you please tell me what the exact commands you put in termux are to make it work?

**Reply by parkerlreed** [reply]
```
pkg i git nodejs npm
git clone https://github.com/novnc/websockify-js
cd websockify-js/websockify
npm i
node websockify.js 15555 adb-ip:adb-port
```

**Reply by CorruptionVR** [reply]
thank you

**Reply by CorruptionVR** [reply]
ok, so i get this error
```
Error: WebSocket connect failed
    at e.onerror (https://tango-adb.github.io/old-demo/_next/static/chunks/pages/_app-b8a1688d9ffda068.js:1:8643)
```
I have made a certificate using openssl-tools, and used the command 
`node websocify.js --cert self.pem --key self.pem 15555 adb-ip:adb-port`(censored the ip and port)
termux tells me that its in wss:// mode. in tango, i put in `wss://adb-ip:adb-port`

**Reply by yume-chan** [reply]
> `wss://adb-ip:adb-port`

Should be `wss://<the machine running websocify>:15555`

**Reply by CorruptionVR** [reply]
Oh, so the ip of the machine?

**Reply by CorruptionVR** [reply]
ok, so now i get this error using `wss://device IP address:15555`
```
Error: WebSocket connect failed
    at WebSocket.<anonymous> (https://tango-adb.github.io/old-demo/_next/static/chunks/pages/_app-b8a1688d9ffda068.js?__WB_REVISION__=b8a1688d9ffda068:1:8643)```

**Reply by yume-chan** [reply]
There can be three devices, one running Tango, one running WebSockify, and one running Android you want to connect. In WebSockify, use the third device's IP address and ADB over WiFi port as the third argument. In Tango, use the second deviece's IP address, the port number is the first argument to WebSockify.

If device number 2 and number 3 are the same device, i.e., WebSockify is running on the Android device you want to connect, then you can use `localhost:<adb port>` as the third argument to WebSockify, and use the device's IP address in Tango.

When device number 1 and number 2 are different, i.e., WebSockify is running on a different machine than Tango is running on, an SSL certificate must be provided to WebSockify. In Tango, use `wss://` instead of `ws://`. If the certificate is not issued by a trusted CA, manually trust the certificate on the device running Tango.

**Reply by greathh** [reply]
@glitch128 Did you solve it? If you solve it, can you share the process?  I get the error is _handler exception: [SSL: SSLV3_ALERT_CERTIFICATE_UNKNOWN] sslv3 alert certificate unknown (_ssl.c:1007)_

**Reply by parkerlreed** [reply]
Yeah, I think the confusion here is that Chrome no longer allows unsecure web socket from an HTTPS domain and generating your own self-signed certificate for WebSockify.js doesn't seem to be as easy as they claim.

**Reply by parkerlreed** [reply]
Okay, I think I got the certificate installed. But I'm running into another issue, but I think it's just because of my connecting device (Quest 2)

![6f74b82ed6a2bc355a91e1a82d1a843c](https://github.com/yume-chan/ya-webadb/assets/841440/aa1fbb9f-ae5d-46e7-b501-92ec4f7dbd47)

![Screenshot_20231028_023230](https://github.com/yume-chan/ya-webadb/assets/841440/f8490007-fdfa-428d-86f3-9a984f8f19bb)
